File: HomePageTest.js
var Browser = require("zombie");
var assert = require("assert");

Browser.visit("http://localhost:8080/", function (e, browser, status) {
    assert.equal(browser.text("title"), "CodeStory - Homepage");
});

Browser.visit("http://localhost:8080/", function (e, browser, status) {
    assert.ok(browser.query("ul li:nth-child(1):contains('dgageot')"));
    assert.ok(browser.query("ul li:nth-child(2):contains('jeanlaurent')"));
    assert.ok(browser.query("ul li:nth-child(3):contains('eric')"));

    assert.ok(browser.query("ul li:nth-child(1):contains('02/12/2012')"));
    assert.ok(browser.query("ul li:nth-child(2):contains('01/12/2012')"));
    assert.ok(browser.query("ul li:nth-child(3):contains('29/11/2012')"));

    assert.ok(browser.query("ul li:nth-child(1):contains('Troisième commit')"));
    assert.ok(browser.query("ul li:nth-child(2):contains('Deuxième commit')"));
    assert.ok(browser.query("ul li:nth-child(3):contains('Premier commit')"));

    assert.ok(browser.query("ul li:nth-child(1) img[src='https://secure.gravatar.com/avatar/f0887bf6175ba40dca795eb37883a8cf']"));
    assert.ok(browser.query("ul li:nth-child(2) img[src='https://secure.gravatar.com/avatar/649d3668d3ba68e75a3441dec9eac26e']"));
    assert.ok(browser.query("ul li:nth-child(3) img[src='https://secure.gravatar.com/avatar/77da98419ae312eb0e322a3dac44a734']"));
});

File: src/test/java/net/gageot/HomePageTest.java
package net.gageot;

import com.google.inject.*;
import net.gageot.codestory.*;
import net.gageot.test.rules.*;
import net.gageot.test.utils.*;
import org.junit.*;

import java.util.*;

import static java.util.Arrays.*;
import static net.gageot.test.rules.ServiceRule.*;
import static org.fest.assertions.Assertions.*;
import static org.mockito.BDDMockito.*;

public class HomePageTest {
	static AllCommits allCommits = mock(AllCommits.class);

	@ClassRule
	public static ServiceRule<CodeStoryServer> server = startWithRandomPort(CodeStoryServer.class, new AbstractModule() {
		@Override
		protected void configure() {
			bind(AllCommits.class).toInstance(allCommits);
		}
	});

	@Test
	public void should_show_homepage() {
		List<Commit> commits = asList( //
				commit("dgageot", "03/01/2012", "Quatrieme commit", "f0887bf6175ba40dca795eb37883a8cf"), //
				commit("jeanlaurent", "03/01/2012", "Troisieme commit", "649d3668d3ba68e75a3441dec9eac26e"), //
				commit("jeanlaurent", "02/01/2012", "Deuxieme commit", "649d3668d3ba68e75a3441dec9eac26e"), //
				commit("eric", "01/01/2012", "Premier commit", "77da98419ae312eb0e322a3dac44a734"));
		given(allCommits.list()).willReturn(commits);

		int exitCode = runZombieJsTest("HomePageTest.coffee");

		assertThat(exitCode).isZero();
	}

	static int runZombieJsTest(String jsFile) {
		return new Shell().execute("./mochalauncher.sh ./src/test/resources/%s %d", jsFile, server.service().getPort());
	}

	static Commit commit(String commiter, String date, String message, String avatarUrl) {
		Commit commit = mock(Commit.class);
		when(commit.getLogin()).thenReturn(commiter);
		when(commit.getDate()).thenReturn(date);
		when(commit.getMessage()).thenReturn(message);
		when(commit.getAvatarUrl()).thenReturn("https://secure.gravatar.com/avatar/" + avatarUrl);
		return commit;
	}
}
